Widom excess chemical potential calculation and RxMC performance improvement

pm-blanco opened the pull request:
Description of changes:
-  `WidomInsertion.measure_excess_chemical_potential()` was replaced by `WidomInsertion.calculate_excess_chemical_potential()` which returns the instantaneous value of the excess chemical potential instead of the accumulated mean and std. The mean value and std of the excess chemical potential must be now be calculated by the user elsewhere. An example of how to do it is given in `samples/widom_insertion.py`.
- `ReactionAlgorithm::do_reaction()` now only computes the potential energy of the old state at the begininng of the reaction loop. This value is updated after each successful reaction trial move.

RudolfWeeber commented:
Contentwise this looks good. However, we use semantic versioning. I.e., as long as the major version (4 in 4.1.2) is the same, we promise not to change the outcome of a simulation script silently.
To make your change non-silent, please rename the python function, i.g., calc_instantaneous_chemical_potential()

jonaslandsgesell commented on samples/widom_insertion.py:
```diff
@@ -118,18 +118,37 @@
+        np.mean(gamma_samples[block * size_block:(block + 1) * size_block]))
+
+mu_ex_mean = -1 * np.log(np.mean(gamma_block)) * temperature
+mu_ex_std = np.sqrt(np.var(gamma_block) / np.mean(gamma_block)) * temperature
```
np.std(..., dof=1) instead of np.sqrt(np.var(...))

jonaslandsgesell replied:
Please add a note that this is due to linearized error propagation. In principle you can do better via not using linearization: mu_ex_Delta =abs( -1 * np.log(np.mean(gamma_block)+np.std(gamma_block, dof =1)) * temperature -(-1 * np.log(np.mean(gamma_block)-np.std(gamma_block, dof =1)) * temperature)

jonaslandsgesell commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -564,20 +564,17 @@ cdef class WidomInsertion(ReactionAlgorithm):
-        the excess chemical potential. The error estimate assumes that
-        your samples are uncorrelated.
-
+        Returns the instantaneous excess chemical potential
```
There is no instantaneous chemical potential (which is always an ensemble average, please find a better name), it would also miss the application of -kT log().

If you change the logic of how this function works, please describe the needed computations in the doc string

jonaslandsgesell commented on testsuite/python/widom_insertion.py:
```diff
@@ -94,19 +94,38 @@ def setUp(self):
+
+        N_block = 16
+        size_block = int(num_samples / N_block)
+        gamma_block = []
```
Please encapsulate the logic in a blocking method

jonaslandsgesell commented:
There is no such thing as an instanteous chemical potential. The chemical potential is per definition an ensemble average. Additionally, the value returned cannot be energy: The application of -kT log is missing (surely because log and the ensemble average do not commute). You could name it compute_exp_of_potential_energy_change_for_excess_chemical_potential_computatiom() which would at be a telling name.

jonaslandsgesell commented on src/core/reaction_methods/WidomInsertion.cpp:
```diff
@@ -29,7 +29,7 @@
 namespace ReactionMethods {
 
-std::pair<double, double> WidomInsertion::measure_excess_chemical_potential(
+double WidomInsertion::calc_instantaneous_chemical_potential(
```
The name is misleading

jonaslandsgesell commented on src/core/reaction_methods/WidomInsertion.cpp:
```diff
@@ -62,15 +62,12 @@ std::pair<double, double> WidomInsertion::measure_excess_chemical_potential(
+
+  const double mu = E_pot_new - E_pot_old;
+
+  return mu;
```
The name is misleading it is a potential energy change not an instanteous chemical potential. By the way: the chemical potential mu also has contributions from the kinetic energy of particles (this is why also an ideal gas has a chemical potential). The Widom method only deals with the excess chemical potential due to interactions. Just stating mu would be misleading.

jonaslandsgesell reviewed (changes requested):
The current naming is misleading, please improve

jonaslandsgesell commented on src/core/reaction_methods/ReactionAlgorithm.cpp:
```diff
@@ -49,9 +49,22 @@ namespace ReactionMethods {
+
+  // calculate potential energy
+  double E_pot_old =
+      calculate_current_potential_energy_of_system(); // only consider potential
```
Naming

jonaslandsgesell commented on src/core/reaction_methods/ReactionAlgorithm.cpp:
```diff
@@ -291,7 +304,7 @@ std::map<int, int> ReactionAlgorithm::save_old_particle_numbers(
 void ReactionAlgorithm::generic_oneway_reaction(
-    SingleReaction &current_reaction) {
+    SingleReaction &current_reaction, double E_pot_old) {
 
```
Why would you submit E_pot old from extern? You can just compute it at the beginning of this function before any change happens. To me reusing old E_pot_old values looks wrong. Please check the logic of your change!

pm-blanco replied:
@jonaslandsgesell  Thank you for your carefull revision of the changes. Let me answer to some of your points:

- I agree with you that the name "instantaneous chemical potential" is missleading. What do you think about renaming it as "particle insertion potential energy"?

- Regarding the reason of submiting E_pot_old from extern. The point is that you only need to compute the energy of the old state the first time that you call the reaction routine. You can avoid to have to compute two times the system energy per reaction by simply updating E_pot_old when a reaction is accepted. Since it is an important change I checked that this implementation does not have any artifact comparing the results of the constant pH tutorial with electrostatics with the current 4.1.4. version of espresso.

jonaslandsgesell replied:
I would be fine with your suggested name.
Right now, I do not see how your change does not change the behavior of the generic one way function. If you perform multiple reactions say n> 2 it has to change behaviour, if e.g. the first reaction is accepted and the energy of the system changes? If not, why?

pm-blanco replied:
Then the value of `E_pot_old` is updated to mach the energy of the new state, see in line 367 in `src/core/reaction_methods/ReactionAlgorithm.cpp`,  `E_pot_old = E_pot_new;`

jonaslandsgesell replied:
You pass by value, in order for E_pot_old to be modified in the outside scope you would need to pass by reference.

jonaslandsgesell replied:
- [x] @RudolfWeeber @pm-blanco  Ideally, we would now write a test which fails if the function signature is such that values are passed by value and not by reference.

jngrad replied:
I guess this could be unit tested in `src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp` by adding a `using ReactionAlgorithm::generic_oneway_reaction;` in the declaration of `ReactionAlgorithmTest` between line 56 and 57, and then adding a new scope at line 232 (between the Monte-Carlo scope and the random number generator scope) with a non-uniform Hamiltonian that would force the move to be accepted.

jonaslandsgesell commented on samples/widom_insertion.py:
```diff
@@ -123,7 +123,7 @@
 for i in range(n_iterations):
 
-    mu_ex = widom.calc_instantaneous_chemical_potential(insertion_reaction_id)
+    E_ins = widom.calc_particle_insertion_potential_energy(insertion_reaction_id)
```
consider E_pot_insertion. The inserted particles also carry kinetic energy which is not reported in this observable

jonaslandsgesell commented on src/core/reaction_methods/WidomInsertion.cpp:
```diff
@@ -63,11 +63,11 @@ double WidomInsertion::calc_instantaneous_chemical_potential(
+  // calculate calc_particle_insertion_potential_energy
 
-  const double mu = E_pot_new - E_pot_old;
+  const double E_ins = E_pot_new - E_pot_old;
```
I think we should avoid abbreviation whenever possible https://medium.com/ltunes/what-is-clean-code-naming-conventions-part-1-426d383eb85d

jonaslandsgesell commented on src/core/reaction_methods/WidomInsertion.hpp:
```diff
@@ -30,7 +30,7 @@ class WidomInsertion : public ReactionAlgorithm {
   WidomInsertion(int seed) : ReactionAlgorithm(seed) {}
   double
-  calc_instantaneous_chemical_potential(SingleReaction &current_reaction);
+  calc_particle_insertion_potential_energy(SingleReaction &current_reaction);
```
Abbreviation for calculate?

pm-blanco replied:
I don't mind changing the nomenclature but is not "calc" as abbreviation for calculate clear enough? I agree that one should try to avoid to abbreviate too much but also I think that too long names can become confusing

jonaslandsgesell replied:
I don't think it s worth to abbreviate words like this

pm-blanco commented:
@jonaslandsgesell  I belive I have resolved all your comments, please take a look and let me know if you find anything else

jonaslandsgesell commented:
Thank you for taking care.

jonaslandsgesell commented on samples/widom_insertion.py:
```diff
@@ -118,18 +118,54 @@
+gamma_mean, gamma_std = do_block_analysis(gamma_samples)   
+
+mu_ex_mean = -1 * np.log(gamma_mean) * temperature
+mu_ex_std = abs(-1 * np.log(gamma_mean) + gamma_std) * temperature - \
```
I would write: 
`#full propagation of error`
`mu_ex_Delta = 0.5*temperature* abs(-np.log(gamma_mean + gamma_std)  - \
    (-np.log(gamma_mean - gamma_std)) )`
where I
* made sure abs goes around everything
* simplified -1*log to -log (I would also do so in line 163)
* added a comment
* pulled out temperature
* moved gamma_std within the argument of the first log
* the 0.5 stems from the standard deviation also just measuring something like a "half width"


Could you please consider to add the function

`compute_excess_chemical_potential_and_error(mean_potential_insertion_energy, standard_deviation_potential_insertion_energy)` into the WidomInsertion class in the .pyx file? Like this your effort of coming up with the correct computations is available for everyone.
You could also consider making the function signature

`compute_excess_chemical_potential_and_error(potential_energy_changes_due_to_insertions, N_block)` 

such that even the blocking analysis is available for everyone and the mean value and standard deviation computation happens inside compute_excess_chemical_potential, given a list potential_energy_changes_due_to_insertions

pm-blanco replied:
I have changed the line as you suggest and I will probably try to make the function in the cython file as well soon (although I do not have the time today).

jonaslandsgesell commented:
This PR would be a good check for your benchmark PR. What is the performance before/after your suggested change, when running e.g. 10 reactions in a row for 1000 times?

pm-blanco commented:
@jonaslandsgesell I did a performance test with the benchmark script of my PR. Performing 1000 MC reactions steps  prior my implementation takes ~ 1.5 s on my laptop. Performing the same number of MC steps with the implementation proposed in this PR takes ~ 0.9 s on my laptop. For a sufficiently large number of reaction steps you gain roughly a factor of 2 in computational time (only in the reaction side, of course).

jonaslandsgesell commented:
Could you please add the unittests which checks if the energy is altered after it is passed in? Then this PR is fine from my side

jngrad commented:
Once the comments by Jonas are resolved, it would be nice to rebase the PR. I find it challenging to follow the changes, because the 19 commits alternate between the Widom refactoring and the `ReactionAlgorithm::do_reaction` for loop optimization (plus a few random merge and formatting commits). Also most commits don't pass CI, so if I select a commit revision and the code doesn't build or doesn't pass the tests on my workstation, I don't know if the issue comes from my build environment or from the PR.

jngrad commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -612,20 +612,59 @@ cdef class WidomInsertion(ReactionAlgorithm):
             deref(self.WidomInsertionPtr).reactions[int(2 * reaction_id)])
+
+    def calculate_excess_chemical_potential(
+            self, particle_insertion_potential_energy_samples, N_blocks=16):
```
I'm not a big fan of long function names and variable names like `particle_insertion_potential_energy_samples`. Wouldn't it be sufficient to call it `energies`? The docstring already explains this argument contains the particle insertion potential energy.

jonaslandsgesell replied:
I am not a fan of too short variable names. For me this would be fine. I prefer longer names over shorter names.

jngrad commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -612,20 +612,59 @@ cdef class WidomInsertion(ReactionAlgorithm):
+                    np.mean(samples[block * size_block:(block + 1) * size_block]))
+
+            sample_mean = np.mean(block_list)
+            sample_std = np.std(block_list)
```
1. by default `np.std()` divides the variance by `n` instead of `n - 1`, shouldn't we use `ddof=1` here?
2. when doing binning analysis, aren't we interested in the standard error of the mean? i.e. `sample_std/np.sqrt(N_blocks)`

jonaslandsgesell replied:
In the Janke publications the formula is provided as suggested by Jean-Noel.
I suppose the derivation of dividing by N_blocks would go by the fact that one computes the variance of the blocking mean

<img src="https://render.githubusercontent.com/render/math?math=%5Cfrac%7B1%7D%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Csum_%7Bi%3D1%7D%5E%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Cmu_%7B%5Cmathrm%7Bblock%7D%2C%20i%7D" alt="\frac{1}{\mathrm{N_blocks}} \sum_{i=1}^{\mathrm{N_blocks}} \mu_{\mathrm{block}, i}" height=32>

i.e.

<img src="https://render.githubusercontent.com/render/math?math=%5Cmathrm%7BVar%7D%5Cleft(%5Cfrac%7B1%7D%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Csum_%7Bi%3D1%7D%5E%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Cmu_%7B%5Cmathrm%7Bblock%7D%2C%20i%7D%5Cright)" alt="\mathrm{Var}(\frac{1}{\mathrm{N_blocks}} \sum_{i=1}^{\mathrm{N_blocks}} \mu_{\mathrm{block}, i})" height=32>

(and since each block is uncorrelated to the next one) we get:

<img src="https://render.githubusercontent.com/render/math?math=%5Cfrac%7B1%7D%7B%5Cmathrm%7BN_blocks%7D%5E2%7D%20%5Csum_%7Bi%3D1%7D%5E%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Cmathrm%7BVar%7D%5Cleft(%5Cmu_%7B%5Cmathrm%7Bblock%7D%2C%20i%7D%5Cright)%20%3D%20%5Cfrac%7B1%7D%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Cmathrm%7BVar%7D%5Cleft(%5Cmu_%7B%5Cmathrm%7Bblock%7D%7D%5Cright)" alt="\frac{1}{\mathrm{N_blocks}^2} \sum_{i=1}^{\mathrm{N_blocks}} \mathrm{Var}(\mu_{\mathrm{block}, i}) = \frac{1}{\mathrm{N_blocks}} \mathrm{Var}(\mu_{\mathrm{block}})" height=35>

so that finally the estimated blocking estimate for the standard deviation of the blocking mean is:

<img src="https://render.githubusercontent.com/render/math?math=%5Csqrt%7B%5Cfrac%7B1%7D%7B%5Cmathrm%7BN_blocks%7D%7D%20%5Cmathrm%7BVar%7D%5Cleft(%5Cmu_%7B%5Cmathrm%7Bblock%7D%7D%5Cright)%7D" alt="\sqrt{\frac{1}{\mathrm{N_blocks}} \mathrm{Var}(\mu_{\mathrm{block}})}" height=45>

jonaslandsgesell replied:
@pm-blanco, to my and Jean-Noels @jngrad understanding you need to return sample_std/np.sqrt(N_blocks) in the blocking analysis for estimating the uncertainty of the blocking mean

pm-blanco replied:
@jonaslandsgesell  solved, I thought that it was encapsulated when one gaves the argument dof=1 to np.std()

jngrad commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -612,20 +612,59 @@ cdef class WidomInsertion(ReactionAlgorithm):
+        def do_block_analysis(samples, N_blocks=16):
+            """
+            Performs a binning analysis of samples. 
+            Divides the samples in N_blocks equispaced blocks
```
```suggestion
            Divides the samples in ``N_blocks`` equispaced blocks
```
The two backticks will format that word in monospace font and label it as source code. My spellchecker software relies on that information to skip variable names and source code samples.

jngrad commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -612,20 +612,59 @@ cdef class WidomInsertion(ReactionAlgorithm):
+        Given a set of samples of the particle insertion potential energy, calculates the excess chemical potential 
+        and its statistical error.
+        Returns excess_chemical_potential_mean, excess_chemical_potential_error   
+        """
```
Please follow the numpydoc + napoleon style ([wiki page](https://github.com/espressomd/espresso/wiki/Documentation#Sphinx)). It *could* look like this:
```suggestion
        """
        Given a time series of the particle insertion potential energy,
        calculates the excess chemical potential and its statistical error.

        Parameters
        ----------
        particle_insertion_potential_energy_samples : array_like of :obj:`float`
            Time series of the particle insertion potential energy.
        N_blocks : :obj:`int`, optional
            Number of bins.

        Returns
        -------
        mean : :obj:`float`
            Mean excess chemical potential.
        error : :obj:`float`
            Standard error of the mean.
        """
```
Do not commit this suggestion directly, as it *does not* describe what your code is currently doing (e.g. the error is not the standard error of the mean).

jngrad commented on samples/widom_insertion.py:
```diff
@@ -118,18 +118,28 @@
+    "excess chemical potential for an ion pair ",
+    mu_ex_mean,
+    " +/- ",
+    mu_ex_Delta)
```
f-strings are more readable (I'm not sure about the floating-point precision we need here, though):
```suggestion
print(
   f"excess chemical potential for an ion pair {mu_ex_mean:.4f} +/- {mu_ex_Delta:.4f}")
```

pm-blanco commented:
@jonaslandsgesell @jngrad  I will try to attend your comments ASAP. Honestly, I have not yet done the unit test because I am unsure on how to implement such test straightforwardly, but I will try to propose something.

jonaslandsgesell commented:
> @jonaslandsgesell @jngrad I will try to attend your comments ASAP. Honestly, I have not yet done the unit test because I am unsure on how to implement such test straightforwardly, but I will try to propose something.

@jngrad , could you please assist in the unittest? It might require refactoring to get something which is easily testable.

pm-blanco commented:
@jonaslandsgesell it took me a while to figure it out how to do it properly, but I managed to implement one unit test in `src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp`. Basicaly I define a simple identity change reaction D <-> E with a reaction constant Gamma = infinity (i.e. completly shifted to products). Then I create one particle D in the system and I call `ReactionAlgorithm::generic_oneway_reaction` providing a fake energy of the system equal to 1. Since it is an ideal system with no interactions and the reaction always will be accepted, after this call the energy should be updated to 0. I checked and the test fails if in `ReactionAlgorithm::generic_oneway_reaction` the energy is provided by value and not by reference and it passes otherwise.
@jngrad I am sorry, but I did not rebase the git history this time. I prefer to practice and learn how to do it properly than mess this PR. Next time I will try to do a more clean PR.  I applied all your other revisions.

pm-blanco commented:
@jngrad it is now failing the CI due to some weird error due to an "unexpected indentation in the docstring" in `\src\python\espressomd\reaction_ensemble.pyx`, any idea why?

jngrad commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -612,20 +612,71 @@ cdef class WidomInsertion(ReactionAlgorithm):
+        -------
+        mean : :obj:`float`
+            Mean excess chemical potential.
+        error : :obj:`float`
+            Standard deviation of the mean.
```
I don't think that's true. If I'm reading the block analysis code correctly, it currently returns the standard error of the mean, not the standard deviation.

jngrad commented on src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp:
```diff
@@ -26,12 +26,12 @@
 #include <boost/test/unit_test.hpp>
 
 #include "reaction_methods/ReactionAlgorithm.hpp"
+#include "reaction_methods/ReactionEnsemble.hpp"
```
The `ReactionEnsemble` header cannot be included it this file, because it's a unit test for the `ReactionAlgorithm` class. Please either rewrite the new test logic to use the `ReactionAlgorithm` class, or move it to the `ReactionEnsemble_test.cpp` file.

jngrad commented on src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp:
```diff
@@ -41,6 +41,13 @@
 #include <utility>
 #include <vector>
 
+#define OBSERVABLES_ENERGY_HPP
+#include "electrostatics_magnetostatics/coulomb.hpp"
+#include "energy.hpp"
+#include "observables/Observable.hpp"
+#include <iostream>
+using namespace std;
```
Please don't write `#define OBSERVABLES_ENERGY_HPP`. This will silence error messages that would otherwise help in debugging linker errors. You might also instantiate the wrong class templates and function templates without knowing it. We also don't use `using namespace std;` because it interferes with the namespace resolution mechanism. We also don't need `<iostream>`. The other includes need to be moved to the top, in the following order of precedence: espresso core headers before `utils` headers before `boost` before `stl`.

jngrad commented on src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp:
```diff
@@ -230,6 +238,68 @@ BOOST_AUTO_TEST_CASE(ReactionAlgorithm_test) {
+  {
+
+// Create an instance of the Reaction ensemble class
+#include "reaction_methods/ReactionEnsemble.hpp"
```
An include statement in the middle of a function, namespace or boost unittest is dangerous. Please remove it.

jngrad commented on src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp:
```diff
@@ -230,6 +238,68 @@ BOOST_AUTO_TEST_CASE(ReactionAlgorithm_test) {
+    test_reaction.charges_of_types[0] = 0;
+    test_reaction.charges_of_types[1] = 0;
+
+    float Gamma = INFINITY; // Reaction completly shifted to product creation
+    ReactionMethods::SingleReaction reaction(Gamma, {type_D}, {1}, {type_E},
+                                             {1});
```
Variable names start with a lowercase to distinguish them from struct/class names. We use `double` for floating-point values, except in CUDA code. We don't use `INFINITY` nor `std::numeric_limits<double>::infinity()` because infinity values are ignored when ESPResSo is compiled in fast-math mode. Would 10^100 be enough? I thought about `std::numeric_limits<double>::max()`, but then if any value is added to it when calculating the probability, there is a risk of overflow.

```suggestion
    auto const gamma = 1e100; // Reaction completely shifted to product creation
    ReactionMethods::SingleReaction reaction(gamma, {type_D}, {1}, {type_E},
                                             {1});
```

jngrad commented on src/core/reaction_methods/tests/ReactionAlgorithm_test.cpp:
```diff
@@ -230,6 +238,68 @@ BOOST_AUTO_TEST_CASE(ReactionAlgorithm_test) {
+    place_particle(0, ref_position);
+    set_particle_type(0, type_D);
+
+    // Add a fake system energy just to check it gets updated
+    double energy = 1.0;
```
```suggestion
    // sentinel value to check energies get updated
    double energy = 1.0;
```

jngrad commented on src/python/espressomd/reaction_ensemble.pyx:
```diff
@@ -612,20 +612,71 @@ cdef class WidomInsertion(ReactionAlgorithm):
+        """
+        Given a set of samples of the particle insertion potential energy,
+        calculates the excess chemical potential and its statistical error.
+        Parameters
+        ----------
+        particle_insertion_potential_energy_samples : array_like of :obj:`float`
+            Samples of the particle insertion potential energy.
+        N_blocks : :obj:`int`, optional
+            Number of bins for binning analysis.
+        Returns
+        -------
+        mean : :obj:`float`
+            Mean excess chemical potential.
```
numpydoc mandates empty lines before sections titles such as `Parameters` and `Returns`, otherwise you get indentation errors in sphinx.

jngrad commented:
For the rebase, it's as simple as
```sh
# backup current state of the branch
git branch Widom+RxMC_fix_backup
# squash all changes
git reset 5bab6bd51e7
```
Then do `git add -p` to select which lines belong to the performance improvement change for the reaction algorithm, commit it, and add the rest to the Widom commit. You can recover the original history at any time with the backed up branch.

pm-blanco committed:
core: Improve efficiency of the reaction ensemble algorithm

Cache the potential energy value. This saves the cost of an energy
calculation per loop when running multiple reactions.

pm-blanco committed:
python: Re-implement Widowm energy calculation

Widom now returns the particle insertion potential energy and
calculates the excess chemical potential afterwards with block
analysis.

pm-blanco commented:
@jngrad  the PR is now blocked because it does not detect that I did the changes you requested, I belive
